Blog - An Introduction to Kubernetes Secrets (#343)
earthly/website: blog/_posts/2023-03-17-kubernetes-secrets.md

kubectl create -n example secret generic postgres-demo \
--from-literal=username=johndoe --from-literal=password=123456
~~~

<div class="wide">
![Creating Secret postgres-demo]({{site.images}}{{page.slug}}/TmXatgQ.png)
</div>

Considering the command above there are few things to note:

- The term `generic` is used to give the secret a generic name. Depending on what you want to use the secret for, you can also have a *docker-registry* or a *tls* secret, used for a docker-registry or as a TLS secret, respectively. You can run  `kubectl create secret --help` so see the available commands for creating a secret.
- The `-from-literal=` is used to specify a key and literal value to insert in secret (`mykey=somevalue`). You can run  `kubectl create secret generic --help` for more information on creating generic secrets.
- There is no need to encrypt the value as it's done automatically.

Run the command below to view the secret:

~~~{.bash caption=">_"}
kubectl get secret -n example
~~~

You can see the secret `postgres-demo` was created with a secret type of **opaque:**

<div class="wide">
![Viewing Secret postgres-demo]({{site.images}}{{page.slug}}/83fPibh.png)
</div>

> In Kubernetes there are various types of secrets, we have:

Reviewer: ```suggestion
In Kubernetes there are various types of secrets; we have:
```

~~~{.bash caption=">_"}
kubectl describe secret postgres-demo -n example
~~~

The output below shows you the secret, the type, and its keys alongside its encrypted values:

<div class="wide">
![Describing Secret(postgres-demo)]({{site.images}}{{page.slug}}/mGG33fw.png)
</div>

### Creating a Secret From a File

You can also create a secret from a file. This file could be a file containing some value.

Using your preferred editor, create two files. A *username.txt* and *password.txt* file. This tutorial uses `nano`.

Type in *janedoe* and *123456789* in the *username.txt* and *password.txt* file respectively:

~~~{.bash caption=">_"}
nano username # Type in *janedoe* and save the file 
nano password # Type in *123456789* and save the file 
~~~

Execute the command below to create this secret:

~~~{.bash caption=">_"}
kubectl create -n example secret generic postgres-demo-0 \
--from-file=./username.txt --from-file=./password.txt
~~~

<div class="wide">
![Creating Secret(postgres-demo-0)]({{site.images}}{{page.slug}}/7GB0UUb.png)
</div>

Run the command below, to see the newly created secret:

~~~{.bash caption=">_"}
kubectl get secret -n example
~~~

> You should now have two secrets - *postgres-demo*  and *postgres-demo-0*, as shown below:

Reviewer: ```suggestion
You should now have two secrets: *postgres-demo*  and *postgres-demo-0*, as shown below:
```

  resources:
    requests:
      storage: 300Mi # Indicates this claim requests only 300Mi 
                     # of storage from a PV
~~~

Run the command below to create the persistent volume and persistent volume claim:

~~~{.bash caption=">_"}
kubectl apply -f pv.yaml -n example
~~~

<div class="wide">
![Creating Persistent Volume and Claim in Kubernetes Cluster]({{site.images}}{{page.slug}}/6IAuhuq.png)
</div>

Confirm the persistent volume and claim are up and running using the commands below:

~~~{.bash caption=">_"}
kubectl get pv 
# gets the persistent volume for your kubernetes cluster
kubectl get pvc -n example 
# gets the persistent volume claim in the *example* namespace
~~~

<div class="wide">
![Verifying Persistent Volume and Claim]({{site.images}}{{page.slug}}/7LGtteC.png)\
</div>

### Using a Kubernetes Secret as an Environment Variable

Up to this point you have created a persistent volume (PV) and a persistent volume claim (PVC), you are now ready to configure a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) and a service to deploy a PostgreSQL database up to Kubernetes.

> The secret **postgres-secret* holds the credentials needed to access the PostgreSQL server, so you will use that secret as an environment variable to configure a statefulSet to deploy a PostgreSQL database.

Reviewer: ```suggestion
The secret **postgres-secret** holds the credentials needed to access the PostgreSQL server, so you will use that secret as an environment variable to configure a statefulSet to deploy a PostgreSQL database.
```

          image: postgres # The image of the postgres database
          imagePullPolicy: "IfNotPresent"
          env:
          - name: POSTGRES_USER
            value: /var/lib/postgresql/secret/postgres_username
          - name: POSTGRES_PASSWORD
            value: /var/lib/postgresql/secret/postgres_password
          ports:
          - containerPort: 5432 # The port number postgres listens on
          volumeMounts:
          - name: data
            mountPath: /var/lib/postgresql/data # Data should be mounted 
            # onto this file path
          - name: secret-volume
            mountPath: /var/lib/postgresql/secret
      volumes:
      - name: data
        persistentVolumeClaim:
                    claimName: postgres-claim # Indicates the postgres 
                    # database should use a PVC called mysql-claim
      - name: secret-volume
        secret:
          secretName: postgres-secret
          items:
          - key: postgres_password
            path: postgres_password
            mode: 511
          - key: postgres_username
            path: postgres_username
            mode: 511
~~~

Here's what we are doing in the code above:

> Define a volume section under the statefulSet Pod specification, with a name *secret-volume*.
> Specify the type of volume (a secret volume type) and which secret it should use (*postgres-secret*).
> Include the keys Kubernetes should watch out for while creating the statefulset (*postgres_password* and *postgres_username*) alongside their paths which are required *postgres_password* and *postgres_username* (these paths could be anything, like *my-postgres-path*).
> Define a mode **511** which is optional.
> Mount the volume inside the containers section using *volumeMounts* , specify the name of the volume we'd like to mount which is *secret-volume* and then the *mountPath* `/secret` which is where the secret will be saved.
> Added a `env` section to get the POSTGRES_USER and POSTGRES_PASSWORD value form the  */var/lib/postgresql/secret/postgres_username* and */var/lib/postgresql/secret/postgres_password* file paths in the container's file system.
> 

Reviewer: ```suggestion
- Define a volume section under the statefulSet Pod specification, with a name *secret-volume*.
- Specify the type of volume (a secret volume type) and which secret it should use (*postgres-secret*).
- Include the keys Kubernetes should watch out for while creating the statefulset (*postgres_password* and *postgres_username*) alongside their paths which are required (these paths could be anything, like *my-postgres-path*).
- Define a mode **511** which is optional.
- Mount the volume inside the containers section using *volumeMounts* , specify the name of the volume we'd like to mount which is *secret-volume* and then the *mountPath* `/secret` which is where the secret will be saved.
- Add an `env` section to get the POSTGRES_USER and POSTGRES_PASSWORD value from the  */var/lib/postgresql/secret/postgres_username* and */var/lib/postgresql/secret/postgres_password* file paths in the container's file system.

```

kubectl apply -f private-pod.yaml -n example
~~~

<div class="wide">
![Creating pod Specification private-pod]({{site.images}}{{page.slug}}/qNc3YY6.png)
</div>

Confirm if your pod is running without errors:

~~~{.bash caption=">_"}
kubectl get pods -n example
~~~

If you have the below output, then your Kubernetes [cluster](/blog/kube-bench) was able to pull the image from your Dockerhub private repository using the `.docker/config.json` file which contains the authentication credentials for your DockerHub registry.

<div class="wide">
![Viewing pod private-pod]({{site.images}}{{page.slug}}/Zd61bEv.png)
</div>

## Conclusion

In this tutorial, you have learned about Kubernetes secrets. You have also learned how to use a secret as an environmental variable and as a volume mount. In addition, you know how to create a secret from a file, from the command line using `kubectl` , and from YAML manifest files.

You then learned how to go through the process of authenticating your Kubernetes cluster to pull an image from your private repository on DockerHub by creating a secret from the `.docker/config.json` file, and pulling the image using the `imagePullSecrets` attribute.

> Now that you have a good knowledge of Kubernetes secrets, you can now implement it in your application and deploy safely to Kubernetes.

Reviewer: ```suggestion
Now that you have a good knowledge of Kubernetes secrets, you can implement it in your application and deploy safely to Kubernetes.
```
